src/CoreWCF.Primitives/src/CoreWCF/Security/DataProtectionSecurityStateEncoder.cs

```diff
@@ -0,0 +1,91 @@
+namespace CoreWCF.Security
+{
+    using System.Text;
+    using System.Security.Cryptography;
+    //using System.Security.Cryptography.
+    using CoreWCF.Runtime;
+    using System;
+
+    public class DataProtectionSecurityStateEncoder : SecurityStateEncoder
+    {
+        byte[] entropy;
+        bool useCurrentUserProtectionScope;
+
+        public DataProtectionSecurityStateEncoder()
+            : this(true)
+        {
+            // empty
+        }
+
+        public DataProtectionSecurityStateEncoder(bool useCurrentUserProtectionScope)
+            : this(useCurrentUserProtectionScope, null)
+        { }
+
+        public DataProtectionSecurityStateEncoder(bool useCurrentUserProtectionScope, byte[] entropy)
+        {
+            this.useCurrentUserProtectionScope = useCurrentUserProtectionScope;
+            if (entropy == null)
+            {
+                this.entropy = null;
+            }
+            else
+            {
+                this.entropy = Fx.AllocateByteArray(entropy.Length);
+                Buffer.BlockCopy(entropy, 0, this.entropy, 0, entropy.Length);
+            }
+        }
+
+        public bool UseCurrentUserProtectionScope
+        {
+            get
+            {
+                return this.useCurrentUserProtectionScope;
+            }
+        }
+
+        public byte[] GetEntropy()
+        {
+            byte[] result = null;
+            if (this.entropy != null)
+            {
+                result = Fx.AllocateByteArray(this.entropy.Length);
+                Buffer.BlockCopy(this.entropy, 0, result, 0, this.entropy.Length);
+            }
+            return result;
+        }
+
+        public override string ToString()
+        {
+            StringBuilder result = new StringBuilder();
+            result.Append(this.GetType().ToString());
+            result.AppendFormat("{0}  UseCurrentUserProtectionScope={1}", Environment.NewLine, this.useCurrentUserProtectionScope);
+            result.AppendFormat("{0}  Entropy Length={1}", Environment.NewLine, (this.entropy == null) ? 0 : this.entropy.Length);
+            return result.ToString();
+        }
+
+        protected internal override byte[] DecodeSecurityState( byte[] data )
+        {
+            try
+            {
+                return ProtectedData.Unprotect(data, this.entropy, (this.useCurrentUserProtectionScope) ? DataProtectionScope.CurrentUser : DataProtectionScope.LocalMachine);
+            }
```

Comment:
ProtectedData only works on Windows. We need to come up with an alternative. This is used to encrypt the claims information after the client has authenticated and sent back to the client as an opaque blob as part of the SCT token. When the client sends the token back, the service can decrypt the blob and get the claims data back to populate into the SecurityContext. The advantage of ProtectedData class is that the security key is the same between restarts of the service. So a service can be updated and restarted and you don't lose your existing sessions (depending on other features so not 100%). We should probably create another implementation of the base class which uses an RSA key generated at service startup. We can then later add the option of this being injected via DI so users can still get the persistent state feature by using an externally managed key (e.g. some cloud keystore mechanism). This is also important for load balancing between multiple multiple servers using a shared security state encryption/decryption key.

Reply:
Can you open an issue for my comment